### DISCUSSION TITLE
Error 404 Not Found while testing govmomi vapi/rest client using simulator

### DISCUSSION THREAD

**Discussion by pragyad3188** [vmware/govmomi#2887] [Community Help] [answered]
I'm trying to create a govmomi/vapi rest client. The sample code is:
```
var insecure bool = true
type RestClient struct {
 url *url.URL
 restClient *rest.Client
}

func (c *RestClient) GetRestClient(ctx context.Context) (*rest.Client, error){
 soapClient := soap.NewClient(c.url, insecure)
 vim25Client, err := vim25.NewClient(ctx, soapClient)
 if err != nil {
  return nil, err
 }
 c.restClient = rest.NewClient(vim25Client)
 err = c.restClient.Login(ctx, c.url.User)
 if err != nil {
  return nil, err
 }
 return c.restClient, nil
}
```
The rest client is created successfully, I tested it by  retrieving categories from vCenter using this client.
To write their unit tests, I'm using govmomi simulator and vapi/simulator.
```
import(
 "context"
 "testing"
 "net/url"
 "fmt"
 "github.com/vmware/govmomi/simulator"
 _ "github.com/vmware/govmomi/vapi/simulator"
)

func TestGetRestClient(t *testing.T){
 ctx := context.Background()
 model := simulator.ESX()
 defer model.Remove()
 err := model.Create()
 if err != nil {
  t.Fatal(err)
 }
 server := model.Service.NewServer()
 defer server.Close()
 url := server.URL
 var client RestClient = RestClient{url: url,}
 _, err := client.GetRestClient(ctx)
 if err != nil {
  t.Errorf("err = %v", err)
  return
 }
}
```
The testcase should pass, but it returns an error 
```
err= POST "http://127.0.0.1:34691/rest/com/vmware/cis/session: 404 Not Found 
```
I'm not able to understand what I'm missing or doing wrong. I've done similar test for creating a govmomi client using simulator and it didn't give me such error.

Can you help me out with this? Thanks in advance!

**Comment by embano1** [comment]
Just glancing over, would be great to see the line where the error occurs. Feel free to take a look at this code/package how to properly construct a client incl. transparent reauth after session expiration: https://github.com/embano1/vsphere/tree/main/client

**Reply by embano1** [reply]
Btw: looks like auth issue on your end as I don‘t see user/pass being correctly set. Again, feel free to take a look at the above package and feel free to use it as it solves common use cases.

**Reply by pragyad3188** [reply]
Thanks for such quick response.
1. The error occurs at line ```err = c.restClient.Login(ctx, c.url.User)```. 
2. I'm not sure if this is an Auth issue because the restclient gets created successfully and I'm able to interact with vCenter to create new tags/retrieve the tags.[I create a ```RestClient``` variable and set its ```url``` field with my IP address, username and password and use this ```RestClient``` variable to call ```GetRestClient()``` function]
The issue occurs while unit testing.

**Reply by pragyad3188** [reply]
@embano1 Can you please take a look at it?

**Reply by embano1** [reply, marked as answer]
You're registering an ESX model, not vCenter: `model := simulator.ESX()`.

Please follow the example I pasted above in the package or other examples and tests used in `govmomi` how to properly use and test the clients.
